piptools/utils.py

```diff
@@ -152,21 +200,12 @@ def _build_direct_reference_best_efforts(ireq: InstallRequirement) -> str:
 
     # If we get here then we have a requirement that supports direct reference.
     # We need to remove the egg if it exists and keep the rest of the fragments.
-    direct_reference = f"{ireq.name.lower()} @ {ireq.link.url_without_fragment}"
-    fragments = []
-
-    # Check if there is any fragment to add to the URI.
-    if ireq.link.subdirectory_fragment:
-        fragments.append(f"subdirectory={ireq.link.subdirectory_fragment}")
-
-    if ireq.link.has_hash:
-        fragments.append(f"{ireq.link.hash_name}={ireq.link.hash}")
-
-    # Then add the fragments into the URI, if any.
-    if fragments:
-        direct_reference += f"#{'&'.join(fragments)}"
-
-    return direct_reference
+    extras = f"[{','.join(xtr for xtr in sorted(ireq.extras))}]" if ireq.extras else ""
```

Comment:
You can simplify this by doing:

```python
extras = f"[{','.join(sorted(ireq.extras))}]" if ireq.extras else ""
```

Reply:
...as long as each item in `extras` is guaranteed to be a string 🙂

Reply:
Should be a string: https://github.com/pypa/pip/blob/main/src/pip/_internal/req/req_install.py#L84

Reply:
Sure, but it's an internal API; it could change without notice. That said, this is probably fine 🤷‍♂️